virtcontainers/device/drivers/vfio.go

```diff
@@ -58,13 +59,29 @@ func (device *VFIODevice) Attach(devReceiver api.DeviceReceiver) error {
 	// Pass all devices in iommu group
 	for _, deviceFile := range deviceFiles {
 
-		//Get bdf of device eg 0000:00:1c.0
-		deviceBDF, err := getBDF(deviceFile.Name())
-		if err != nil {
-			return err
+		var deviceBDF string
+		var deviceSysfsDev string
+		var err error
+
+		vfioDeviceType := getVFIODeviceType(deviceFile.Name())
+		if (vfioDeviceType == VFIODevice_TYPE_VFIO_NORMAL) {
+			//Get bdf of device eg 0000:00:1c.0
+			deviceBDF = getBDF(deviceFile.Name())
+			deviceSysfsDev = ""
+		} else if (vfioDeviceType == VFIODevice_TYPE_VFIO_MDEV){
+			//Get sysfsdev of device eg f79944e4-5a3d-11e8-99ce-479cbab002e4
+			sysfsDevStr := filepath.Join(iommuDevicesPath, deviceFile.Name())
+			deviceSysfsDev, err = getSysfsDev(sysfsDevStr)
+			if err != nil {
+				return err
+			}
+			deviceBDF = ""
+		} else {
+			return  fmt.Errorf("Incorrect number of tokens found while parsing vfio device type: %s", deviceFile.Name())
```

Comment:
It looks like this block could become a new function. Something like:

```
func getVFIODetails(deviceFileName string) (deviceBDF, deviceSysfsDev string)
```

That would allow you to add a more "fine-grained" unit test for this change.